### PR TITLE
feat: to TensorFlow RaggedTensor

### PR DESCRIPTION
**Pull request by maxymnaumchyk**

### CODE REVIEW

**Review by ianna — COMMENTED** [review]
@maxymnaumchyk - good start! Please, run `pre-commit run —all` locally before pushing.

File: tests/test_3210_to_raggedtensor_from_raggedtensor.py [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+starts2 = ak.index.Index64(np.array([0, 3]))
+stops2 = ak.index.Index64(np.array([3, 5]))
+
+array = np.arange(2 * 3 * 5).reshape(2, 3, 5)
+content2 = ak.contents.NumpyArray(array.reshape(-1))
+inneroffsets = ak.index.Index64(np.array([0, 5, 10, 15, 20, 25, 30]))
+outeroffsets = ak.index.Index64(np.array([0, 3, 6]))
+
+
+def convert_to_raggedtensor():
```
Comment (ianna) [inline tests/test_3210_to_raggedtensor_from_raggedtensor.py]:
```suggestion
def test_convert_to_raggedtensor():
```

File: tests/test_3210_to_raggedtensor_from_raggedtensor.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+    assert to_raggedtensor(array9).to_list() == [[3, 1, 4, 1, 9, 2, 6]]
+
+
+np_array1 = np.array([1.1, 2.2, 3.3, 4.4, 5.5], dtype=np.float32)
+
+offsets1 = ak.index.Index64(np.array([0, 2, 3, 3, 5]))
+content1 = ak.contents.NumpyArray(np_array1)
+
+
+def convert_from_raggedtensor():
```
Comment (ianna) [inline tests/test_3210_to_raggedtensor_from_raggedtensor.py]:
```suggestion
def test_convert_from_raggedtensor():
```

**Review by ianna — CHANGES_REQUESTED** [review]
@maxymnaumchyk - Great job! The implementation is fine. However, the tests do not run in the CI because `pytest` expects the test names starting with `test_`. Please, check the suggestions.

File: tests/test_3210_to_raggedtensor_from_raggedtensor.py (hunk 1 above)
Comment (maxymnaumchyk) [reply tests/test_3210_to_raggedtensor_from_raggedtensor.py]: Thanks for pointing that out! I didn't notice the tests weren't running.

**Review by ianna — APPROVED** [review]
@maxymnaumchyk - Great, thanks! All tests pass:
```python
tests/test_3210_to_raggedtensor_from_raggedtensor.py::test_convert_to_raggedtensor PASSED [ 99%]
tests/test_3210_to_raggedtensor_from_raggedtensor.py::test_convert_from_raggedtensor PASSED [100%]
```
@jpivarski - if you don't have any objections I'll merge it. Thanks!

File: src/awkward/operations/ak_to_raggedtensor.py [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+def _recursive_call(layout, offsets_arr):
+    try:
+        # change all the possible layout types to ListOffsetArray
+        if isinstance(layout, ak.contents.listarray.ListArray):
+            layout = layout.to_ListOffsetArray64()
+        elif isinstance(layout, ak.contents.regulararray.RegularArray):
+            layout = layout.to_ListOffsetArray64()
+        elif (not isinstance(layout, ak.contents.listoffsetarray.ListOffsetArray)) and (
+            not isinstance(layout, ak.contents.numpyarray.NumpyArray)
+        ):
```
Comment (jpivarski) [inline src/awkward/operations/ak_to_raggedtensor.py]: `isinstance` takes a tuple of classes to check for, which simplifies this expression.

```suggestion
        elif not isinstance(
            layout,
            (
                ak.contents.listoffsetarray.ListOffsetArray,
                ak.contents.numpyarray.NumpyArray,
            ),
        ):
```

File: src/awkward/operations/ak_to_raggedtensor.py [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+
+def _impl(array):
+    try:
+        import tensorflow as tf
+    except ImportError as err:
+        raise ImportError(
+            """install the 'tensorflow' package with:
+
+        pip install tensorflow"""
+        ) from err
```
Comment (jpivarski) [inline src/awkward/operations/ak_to_raggedtensor.py]:
```suggestion
        raise ImportError(
            """to use ak.to_raggedtensor, you must install the 'tensorflow' package with:

        pip install tensorflow
or
        conda install tensorflow"""
        ) from err
```

**Review by jpivarski — APPROVED** [review]
It looks good! It's well implemented and well tested. I made two suggestions.

File: src/awkward/operations/ak_to_raggedtensor.py (hunk 3 above)
Comment (maxymnaumchyk) [reply src/awkward/operations/ak_to_raggedtensor.py]: thanks for the comment! It looks nicer indeed.

**Comment by jpivarski** [comment]
@all-contributors please add @maxymnaumchyk for code

**Comment by jpivarski** [comment]
Just a note: we'll need to get these new functions into the documentation. It needs to go into https://github.com/scikit-hep/awkward/blob/main/docs/reference/toctree.txt, as a new section

```
.. toctree::
    :caption: Converting ragged arrays

    generated/ak.from_raggedtensor
    generated/ak.to_raggedtensor
```

under the "Converting rectangular arrays" section and above the "Reading and writing files" section.

That can be a quick little PR, and when the documentation-generating Action is done, there should be a "View Deployment" button that will let you see what the new documentation looks like. You should find the new documentation on the left side-bar of https://awkward-array.org/doc/main/reference/index.html .

**Pull request merged**